src/persist-client/src/restore.rs

```diff
@@ -0,0 +1,89 @@
+//! See documentation on [[restore_blob]].
+
+use crate::internal::encoding::UntypedState;
+use crate::internal::paths::BlobKey;
+use crate::internal::state::State;
+use crate::internal::state_diff::{StateDiff, StateFieldValDiff};
+use crate::internal::state_versions::StateVersions;
+use crate::ShardId;
+use anyhow::anyhow;
+use mz_persist::location::Blob;
+use tracing::info;
+
+/// Attempt to restore all the blobs referenced by the current state in consensus.
+/// Returns a list of blobs that were not possible to restore.
+pub(crate) async fn restore_blob(
+    versions: &StateVersions,
+    blob: &(dyn Blob + Send + Sync),
+    build_version: &semver::Version,
+    shard_id: ShardId,
+) -> anyhow::Result<Vec<BlobKey>> {
+    let diffs = versions.fetch_all_live_diffs(&shard_id).await;
+    let Some(first_live_seqno) = diffs.0.first().map(|d| d.seqno) else {
+        info!("No diffs for shard {shard_id}.");
+        return Ok(vec![]);
+    };
+
+    fn after<A>(diff: StateFieldValDiff<A>) -> Option<A> {
+        match diff {
+            StateFieldValDiff::Insert(a) => Some(a),
+            StateFieldValDiff::Update(_, a) => Some(a),
+            StateFieldValDiff::Delete(_) => None,
+        }
+    }
+
+    let mut not_restored = vec![];
+    let mut check_restored = |key: &BlobKey, result: Result<(), _>| {
+        if result.is_err() {
+            not_restored.push(key.clone());
+        }
+    };
+
+    for diff in diffs.0 {
+        let diff: StateDiff<u64> = StateDiff::decode(build_version, diff.data);
+        for rollup in diff.rollups {
+            // We never actually reference rollups from before the first live diff.
+            if rollup.key < first_live_seqno {
+                continue;
+            }
+            if let Some(value) = after(rollup.val) {
+                let key = value.key.complete(&shard_id);
+                check_restored(&key, blob.restore(&key).await);
+                // Elsewhere, we restore any state referenced in live diffs... but we also
+                // need to restore everything referenced in that first rollup.
+                if rollup.key != first_live_seqno {
+                    continue;
+                }
+                let rollup_bytes = blob
+                    .get(&key)
+                    .await?
+                    .ok_or_else(|| anyhow!("fetching just-restored rollup"))?;
```

Comment:
If versioning is not enabled for the bucket, this anyhow is hit on restore. By that time, it is too late to correct the problem and the ship has sailed.

I think environmentd should check the bucket and refuse to start if versioning is not enabled.

Reply:
I've updated the code to skip trying to fetch the rollup if the restore failed. That way, we will continue to make progress through the rest of the restores instead of exiting early.

Since versioning can be enabled & disabled, it's hard to guarantee that it's on when we need it... it might be disabled just after our check. But it would be better to have more safety checking here... I'll take this as a followup.